pkg/aws/ec2.go

```diff
@@ -0,0 +1,62 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/smithy-go"
+)
+
+type EC2Client interface {
+	DescribeImages(ctx context.Context, params *ec2.DescribeImagesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeImagesOutput, error)
+	DescribeKeyPairs(ctx context.Context, params *ec2.DescribeKeyPairsInput, optFns ...func(*ec2.Options)) (*ec2.DescribeKeyPairsOutput, error)
+	CreateKeyPair(ctx context.Context, params *ec2.CreateKeyPairInput, optFns ...func(*ec2.Options)) (*ec2.CreateKeyPairOutput, error)
+}
+
+func NewEC2Client(config aws.Config) *ec2.Client {
+	return ec2.NewFromConfig(config)
+}
+
+func (a *Client) ImageExists(ctx context.Context, imageID string) (bool, error) {
+	params := &ec2.DescribeImagesInput{
+		ImageIds: []string{imageID},
+	}
+	_, err := a.ec2.DescribeImages(ctx, params)
+	if err == nil {
+		return true, nil
+	}
+
+	var apiErr smithy.APIError
+	if errors.As(err, &apiErr) && apiErr.ErrorCode() == "400" {
+		return false, nil
+	}
+	return false, fmt.Errorf("aws describe image [%s]: %v", imageID, err)
+}
+
+func (a *Client) KeyPairExists(ctx context.Context, keyName string) (bool, error) {
```

Comment:
I'm not sure a key name being on the target means the key pair exists/is valid. Could you elaborate on expectations here? My train of thought is that key names could be different between my machine and the host machine but the key pairs are still valid. 

Additionally, why does a client perform this check? I would think the client merely provides behavior for interacting with the target but the actual "does this exist on the target" logic would exist elsewhere. Concretely, a client could provide a `RetrieveKeysByName()` method and some other (probably a validating construct) uses that to retrieve the keys and then performs the necessary checks to ensure the key pair is valid (i.e. the business logic). Why is this important? Your validator is now isolatable because you can use an interface to mock up the `RetrieveKeysByName()` and test the business logic very easily.

Reply:
If we go this route, I would suggest we make sure the client doesn't return any aws sdk specific type.

Reply:
this is amazon ec2 key pairs: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html

Reply:
#### EC2 key pairs

I read the function name as checking if a key pair exists hence felt the behavior was wrong. Looking at the replies and docs you linked, I think the method name is wrong. We're only trying to provide a guarentee that the key name they supplied exists on the EC2 instance so the method should reflect that: `KeyNameExists()`. Does that resonate?

Off the back of that, all the methods seem to be EC2 specific which makes me think the struct name isn't specific enough - could we rename it to `EC2Client`?

--- 

Re behavior splitting among constructs: I empathize with your sentiment - 👍🏻. I don't know if its the right thing to do because "[doesn't contain] much validation logic" > 0 which means this construct both interfaces with EC2 APIs and validates stuff hence could be violating the SRP. On the flip side, these could just be framed as convenience methods on the EC2 API so I commit!

Reply:
1. I feel we should stick to KeyPairExists. We check the aws ec2 key pair existence by the key name.
2. For now it is ec2 specific but I want this struct to be general enough to handle other aws services.